File: Php/integration/Sharing/ResultSharerImpl.php
<?php
namespace integration\Sharing{

	use entities\Cashgame;
	use Infrastructure\Data\Interfaces\SharingStorage;
	use Infrastructure\Data\Interfaces\UserStorage;
	use entities\CashgameResult;
	use Domain\Classes\User;

	class ResultSharerImpl implements ResultSharer{

		private $sharingStorage;
		private $userStorage;
		private $socialServiceFactory;

		public function __construct(SharingStorage $sharingStorage,
									UserStorage $userStorage,
									SocialServiceFactory $socialServiceFactory){
			$this->sharingStorage = $sharingStorage;
			$this->userStorage = $userStorage;
			$this->socialServiceFactory = $socialServiceFactory;
		}

		public function shareResult(Cashgame $cashgame){
			$results = $cashgame->getResults();
			foreach($results as $result){
				$this->shareSingleResult($result);
			}
		}

		public function shareSingleResult(CashgameResult $result){
			$user = $this->getUser($result);
			$services = $this->sharingStorage->getServices($user);
			foreach($services as $serviceName){
				$this->shareToService($serviceName, $user, $result);
			}
		}

		public function shareToService($serviceName, User $user, CashgameResult $result){
			$service = $this->socialServiceFactory->makeSocialService($serviceName);
			$service->shareResult($user, $result->getWinnings());
		}

		private function getUser(CashgameResult $result){
			return $this->userStorage->getUserByName($result->getPlayer()->getUserName());
		}

	}

}

File: Infrastructure/Services/ResultSharerImpl.cs
using Core.Classes;
using Core.Repositories;
using Core.Services;

namespace Infrastructure.Services{

	public class ResultSharerImpl : IResultSharer
    {
	    private readonly ISharingRepository _sharingRepository;
	    private readonly IUserRepository _userRepository;
	    private readonly ISocialServiceProvider _socialServiceFactory;

	    public ResultSharerImpl(
            ISharingRepository sharingRepository,
			IUserRepository userRepository,
			ISocialServiceProvider socialServiceFactory)
	    {
	        _sharingRepository = sharingRepository;
	        _userRepository = userRepository;
	        _socialServiceFactory = socialServiceFactory;
	    }

	    public void ShareResult(Cashgame cashgame){
			foreach(var result in cashgame.Results){
				ShareSingleResult(result);
			}
		}

		private void ShareSingleResult(CashgameResult result){
			var user = GetUser(result);
			var services = _sharingRepository.GetServices(user);
			foreach(var service in services){
				ShareToService(service, user, result);
			}
		}

		public void ShareToService(string serviceName, User user, CashgameResult result){
			var service = _socialServiceFactory.Get(serviceName);
			service.ShareResult(user, result.Winnings);
		}

		private User GetUser(CashgameResult result){
			return _userRepository.GetById(result.Player.UserId);
		}

	}

}